api/app/fireweather_bot/hourly_actuals.py

```diff
@@ -0,0 +1,119 @@
+    """
+    # go from pandas to python native:
+    python_date = panda_datetime.to_pydatetime()
+    # we want to work in utc:
+    python_date = python_date.replace(tzinfo=timezone.utc)
+    # but alse need to adjust our time, because it's in PST:
+    python_date = python_date - timedelta(hours=app.time_utils.PST_UTC_OFFSET)
+    return python_date
+
+
+class HourlyActualsBot(BaseBot):
+    """ Bot that downloads the hourly actuals from the wildfire website and stores it in a database. """
+
+    def __init__(self):
+        self.now = app.time_utils.get_pst_now()
+
+    def _get_start_date(self) -> int:
+        """ Return time N hour ago. E.g. if it's 17h15 now, we'd get YYYYMMDD16. The intention is that
+        this bot runs every hour, so if we ask for everything from an hour back, we should be fine.
+        However, just to be on the safe side, we're asking for the last three hours - just in case there
+        was a station that came in late, or if for whatever reason we missed a run.
+        """
+        hour_ago = self.now - timedelta(hours=3)
+        return int(hour_ago.strftime('%Y%m%d%H'))
+
+    def _get_end_date(self) -> int:
+        """ Return now. E.g. if it's 17h15 now, we'd get YYYYMMDD17 """
+        return int(self.now.strftime('%Y%m%d%H'))
+
+    def process_csv(self, filename: str):
+        with open(filename, 'r') as csv_file:
+            data_df = pd.read_csv(csv_file)
+        station_codes = get_station_names_to_codes()
+        # replace 'display_name' column (station name) in df with station_id
+        # and rename the column appropriately
+        data_df['display_name'].replace(station_codes, inplace=True)
+        data_df.rename(columns={'display_name': 'station_code'}, inplace=True)
+
+        # weather_date is formatted yyyymmddhh as an int - need to reformat it as a Timestamp
+        dates = pd.to_datetime(data_df['weather_date'], format='%Y%m%d%H')
+        data_df['weather_date'] = dates
+        # write to database using _session's engine
+        session = app.db.database.get_session()
+        # write the data_df to the database one row at a time, so that if data_df contains >=1 rows that are
+        # duplicates of what is already in the db, the write won't fail for the unique rows
+        # NOTE: iterating over the data_df one Series (row) at a time is ugly, but until pandas is updated
+        # with a fix, this is the easiest work-around.
+        # See https://github.com/pandas-dev/pandas/issues/15988
+        # pylint: disable=unused-variable
+        for index, row in data_df.iterrows():
+            try:
+                # Go from pandas to a python dict.
+                data = row.to_dict()
+                # Fix the timestamp, make it be tz aware.
+                # Note: There must be a more "pandas" way of doing this, but I don't know how.
+                data['weather_date'] = _fix_pandas_datetime(data['weather_date'])
```

Comment:
I believe it's the map() function you're looking for - https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.map.html

Reply:
@andrea-williams my problem is getting a pandas datetime with a timezone - I don't know how to get pandas to associate a timezone with something, and have it go all the way through to that final dictionary.